internal/topology/check/compatibility.go

```diff
@@ -54,3 +60,96 @@ func ObjectsAreInTheSameNamespace(current, desired client.Object) error {
 	}
 	return nil
 }
+
+// TemplatesAreCompatible checks if two referenced objects are compatible, meaning that
+// they are of the same GroupKind and in the same namespace.
+func TemplatesAreCompatible(current, desired clusterv1.LocalObjectTemplate) error {
+	currentGK := current.Ref.GetObjectKind().GroupVersionKind().GroupKind()
+	desiredGK := desired.Ref.GetObjectKind().GroupVersionKind().GroupKind()
+
+	if currentGK.String() != desiredGK.String() {
+		return errors.Errorf("invalid operation: it is not possible to change the GroupKind of %s/%s from %s to %s",
+			current.Ref.GetObjectKind().GroupVersionKind(), current.Ref.Name, currentGK, desiredGK)
+	}
+	return TemplatesAreInTheSameNamespace(current, desired)
+}
+
+// TemplatesAreInTheSameNamespace checks if two referenced objects are in the same namespace.
+func TemplatesAreInTheSameNamespace(current, desired clusterv1.LocalObjectTemplate) error {
+	// NOTE: this should never happen (webhooks prevent it), but checking for extra safety.
+	if current.Ref.Namespace != desired.Ref.Namespace {
+		return errors.Errorf("invalid operation: it is not possible to change the namespace of %s/%s from %s to %s",
+			current.Ref.GetObjectKind().GroupVersionKind(), current.Ref.Name, current.Ref.Namespace, desired.Ref.Namespace)
+	}
+	return nil
+}
+
+// TemplateIsValid ensures the template is in the correct namespace, has no nil references, and has a valid Kind and GroupVersion.
+func TemplateIsValid(in *clusterv1.LocalObjectTemplate, namespace string, pathPrefix *field.Path) field.ErrorList {
```

Comment:
Question - is it appropriate to return the field.ErrorList here? It breaks with the expected signature in the rest of the package, but changing it could change the output of the webhook to something less useful.

Reply:
Hm on one side I think it would be good to use regular errors and the field errors only in webhooks, on the other side we would lose a lot of information that way.

But looks like the field errors also implement the regular error interface.

Maybe we should store the field errors in a regular array of errors and then use `kerrors.NewAggregate` when we want to return multiple? (i.e. the return parameter would be just `error`).

Reply:
If we return a kerrors.Aggregate (as an error) is there a way to turn it into a []field.Error later on?

Reply:
I would guess no easy way, do we have to convert it back? (aka does it make a difference in the returned error messages)

Not sure if it's possible with the aggregated list of errors, if it would be just one field.Error it should be possible with Go's `errors.As`.

Reply:
>I would guess no easy way, do we have to convert it back? (aka does it make a difference in the returned error messages)

The motivation for using kerrors.Aggregate would be to eventually be able to return a  field.ErrorList in the main webhook right? How can we add those errors into the list? Or are you saying we should stop using field.Error in the webhooks altogether?

Reply:
Spoke with @sbueringer on this. The webhook is returning a StatusError which is created by wrapping the field.ErrorList. There's no obvious way to return an aggregate in a type that conforms to the error {Error() string} interface which we'll be able to convert to an error.ErrorList afterward.

The two choices I see are:
1) Do all complicated checks which may require aggregation of errors in the webhooks package e.g. ValidateAllVariables
2) Change the public functions in the checks package to return a field.ErrorList to make them consistent.

IMO the best solution here is to directly return a field.ErrorList from each function in the check package and modify the current usage in the controller.

Reply:
:+1: for option 2. I think otherwise we use the additional infos from field.Error

Reply:
+1 to option 2, given that at the end all the checks are verifying one field or the other in the objects.
Note. when the func is returning single error, it should return *field.Error instead of field.ErrorList

Reply:
@fabriziopandini @sbueringer I've updated so the error returns are consistent now, including for the existing Object methods - WDYT